### PR TITLE
Chores grooming

### PR DESCRIPTION
**Pull request by pavelkrolevets**

### CODE REVIEW

File: integration_test/integration_test.go [hunk 1]
```diff
@@ -123,32 +142,37 @@ func TestHappyFlow(t *testing.T) {
...
-		srv9 := CreateServer(t, 9)
-		ops[9] = client.Operator{srv9.srv.URL, 9, &srv9.privKey.PublicKey}
-		srv10 := CreateServer(t, 10)
-		ops[10] = client.Operator{srv10.srv.URL, 10, &srv10.privKey.PublicKey}
-		srv11 := CreateServer(t, 11)
-		ops[11] = client.Operator{srv11.srv.URL, 11, &srv11.privKey.PublicKey}
-		srv12 := CreateServer(t, 12)
-		ops[12] = client.Operator{srv12.srv.URL, 12, &srv12.privKey.PublicKey}
-		clnt := client.New(ops)
+		ops := make(map[uint64]initiator.Operator)
```
[the remark this answers is not in the document]
Comment (pavelkrolevets) [reply integration_test/integration_test.go]: understood. Added negative and one positive threshold test to integration_test

File: pkgs/initiator/initiator.go [hunk 2]
```diff
@@ -0,0 +1,790 @@
...
+}
+
+func (c *Initiator) NewID() [24]byte { // random ID for each new DKG initiation
+	var id [24]byte
+	b := uuid.New()
+	copy(id[:12], b[:])
+	b = uuid.New()
+	copy(id[12:], b[:])
+	return id
+}
```
Comment (y0sher) [inline pkgs/initiator/initiator.go:744]: I moved this function to crypto or util, it doesn't need anything from `Initiator` so doesn't need to be a part of it.

File: pkgs/initiator/initiator.go [hunk 3]
```diff
@@ -0,0 +1,790 @@
...
+			return nil, err
+		}
+		// Verify that incoming messages have valid DKG ceremony ID
+		if !bytes.Equal(id[:], tsp.Message.Identifier[:]) {
+			return nil, fmt.Errorf("incoming message has wrong ID, aborting... operator %d, msg ID %x", tsp.Signer, tsp.Message.Identifier[:])
+		}
+		// Verification operator signatures
+		if err := c.VerifyFunc(tsp.Signer, signedBytes, tsp.Signature); err != nil {
+			return nil, err
+		}
```
Comment (y0sher) [inline pkgs/initiator/initiator.go:330]: I moved verification out of this function to VerifyAll, it seems  now we verify twice

File: pkgs/dkg/drand.go [hunk 4]
```diff
@@ -512,3 +499,15 @@ func (o *LocalOwner) broadcastError(err error) {
...
+func (o *LocalOwner) VerifyInitiatorMessage(msg []byte, sig []byte) error {
+	pubKey, err := crypto.EncodePublicKey(o.InitiatorPublicKey)
+	if err != nil {
+		return err
+	}
+	if err := crypto.VerifyRSA(o.InitiatorPublicKey, msg, sig); err != nil {
+		return fmt.Errorf("failed to verify a message from initiator: %x", pubKey)
+	}
+	o.Logger.Info("Successfully verified initiator message signature", zap.Uint64("from", o.ID))
+	return nil
```
Comment (y0sher) [inline pkgs/dkg/drand.go:512]: `VerifyInitiatorMessage` is used only in `state` hence it doesn't make sense to have it here, not the function and not the public key itself.

File: pkgs/dkg/drand.go [hunk 5]
```diff
@@ -102,44 +101,47 @@ type LocalOwner struct {
...
-	OpPrivKey  *rsa.PrivateKey
-	Owner      [20]byte
-	Nonce      uint64
+	Logger             *zap.Logger
+	ID                 uint64
+	BroadcastF         func([]byte) error
+	Suite              pairing.Suite
+	VerifyFunc         func(id uint64, msg, sig []byte) error
+	SignFunc           func([]byte) ([]byte, error)
+	OpPrivKey          *rsa.PrivateKey
```
Comment (y0sher) [inline pkgs/dkg/drand.go:117]: If we pass in SignFunc we don't need OpPrivKey, or vice verca, at first I made it a point not to pass in the key and just use the sign function.

File: pkgs/dkg/drand.go [hunk 6]
```diff
@@ -102,44 +101,47 @@ type LocalOwner struct {
 	VerifyFunc func(id uint64, msg, sig []byte) error
 	SignFunc   func([]byte) ([]byte, error)
 
-	Owner common.Address
-	Nonce uint64
-	done  chan struct{}
+	Owner              common.Address
+	Nonce              uint64
+	done               chan struct{}
+	InitiatorPublicKey *rsa.PublicKey
```
Comment (y0sher) [inline pkgs/dkg/drand.go:107]: this can be part of the wrapper that we have in `state` for the local owner, since the usage is only there it doesn't make sense to put it in here.

**Review by y0sher — CHANGES_REQUESTED** [review]
see my comments

File: pkgs/initiator/initiator.go (hunk 2 above)
Comment (pavelkrolevets) [reply pkgs/initiator/initiator.go:744]: yeah, I saw it at crypto, but got lost after merge, Ill move it to crypto again

Comment (pavelkrolevets) [reply pkgs/initiator/initiator.go]: moved to crypto

File: pkgs/initiator/initiator.go (hunk 3 above)
Comment (pavelkrolevets) [reply pkgs/initiator/initiator.go]: removed

File: pkgs/dkg/drand.go (hunk 4 above)
Comment (pavelkrolevets) [reply pkgs/dkg/drand.go:512]: I ill move it to instWrapper, as you suggested.

File: pkgs/dkg/drand.go (hunk 6 above)
Comment (pavelkrolevets) [reply pkgs/dkg/drand.go]: moved to instWrapper at state

File: pkgs/dkg/drand.go (hunk 5 above)
Comment (pavelkrolevets) [reply pkgs/dkg/drand.go]: we use it also to encrypt the share 
https://github.com/bloxapp/ssv-dkg/blob/chores_grooming/pkgs/dkg/drand.go#L250

Comment (pavelkrolevets) [reply pkgs/dkg/drand.go]: I can add EncryptFunc instead?

Comment (pavelkrolevets) [reply pkgs/dkg/drand.go]: Added EncryptFunc, DecryptFunc, and rsa pub instead of holding private at each instance

**Review by y0sher — APPROVED** [review]
Great job!
@moshe-blox @andrew-blox please take final look so we can merge

File: pkgs/crypto/crypto.go [hunk 7]
```diff
@@ -442,3 +445,55 @@ func VerifyPartialSigs(sigShares map[uint64]*bls.Sign, sharePks map[uint64]*bls.
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file: %w", err)
+	}
+
+	operatorKeyByte, err := base64.StdEncoding.DecodeString(string(data))
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode(operatorKeyByte)
+	// TODO: resolve deprecation https://github.com/golang/go/issues/8860
```
Comment (andrew-blox) [inline pkgs/crypto/crypto.go:474]: Should we resolve this?

File: pkgs/crypto/crypto.go [hunk 8]
```diff
@@ -442,3 +445,55 @@ func VerifyPartialSigs(sigShares map[uint64]*bls.Sign, sharePks map[uint64]*bls.
...
+	enc := x509.IsEncryptedPEMBlock(block) //nolint
+	b := block.Bytes
+	if enc {
+		var err error
+		// TODO: resolve deprecation https://github.com/golang/go/issues/8860
```
Comment (andrew-blox) [inline pkgs/crypto/crypto.go:479]: Should we resolve this?

**Review by andrew-blox — APPROVED** [review]

**Review by y0sher — APPROVED** [review]

File: cli/initiator/initiator.go [hunk 9]
```diff
@@ -29,138 +30,209 @@ func init() {
...
+			logger.Fatal("😥 Please provide a valid fork name: mainnet, prater, or now_test_network")
 		}
 		owner := viper.GetString("owner")
 		if owner == "" {
-			logger.Fatal("failed to get owner address flag value", zap.Error(err))
+			logger.Fatal("😥 Failed to get owner address flag value: ", zap.Error(err))
 		}
 		nonce := viper.GetUint64("nonce")
-		withdrawPubKey, err := hex.DecodeString(withdrawAddr)
+		withdrawPubKey := common.HexToAddress(withdrawAddr).Bytes()
```
Comment (moshe-blox) [inline cli/initiator/initiator.go:215]: This is dangerous for user input because it fails silently, so you can pass just a "0" and this would not fail.

WDYT about making our own parsing func that would return an error if it's too long, too short or not a valid hex?

File: pkgs/operator/operator.go [hunk 10]
```diff
@@ -132,10 +134,14 @@ func New(key *rsa.PrivateKey) *Server {
 }
 
 func (s *Server) Start(port uint16) error {
-	s.Logger.Infof("server listening for incoming requests on port %d", port)
 	srv := &http.Server{Addr: fmt.Sprintf(":%v", port), Handler: s.Router}
```
Comment (moshe-blox) [inline pkgs/operator/operator.go:137]: Not part of this PR, but it'd be nice if we can add timeouts to the `http.Server`

File: pkgs/operator/state.go [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+		PrivateKey:       pv,
+	}
+}
+
+func (s *Switch) InitInstance(reqID [24]byte, initMsg *wire.Transport, initiatorSignature []byte) ([]byte, error) {
+	logger := s.Logger.With(zap.String("reqid", hex.EncodeToString(reqID[:])))
+	logger.Info("🚀 Initializing DKG instance")
+	init := &wire.Init{}
+	if err := init.UnmarshalSSZ(initMsg.Data); err != nil {
+		return nil, err
```
Comment (moshe-blox) [inline pkgs/operator/state.go:173]: We should return error wrappings (as you did below in `VerifyRSA`) here and in the rest of the code, so we can tell where exactly things failed in retrospect.

**Review by moshe-blox — CHANGES_REQUESTED** [review]
Looks good 👍

Left a few comments, although did not finish reviewing yet.

File: pkgs/operator/operator.go (hunk 10 above)
Comment (pavelkrolevets) [reply pkgs/operator/operator.go:137]: We have Chi limiting requests to routes, to prevent DDOS, if I understood you correctly

File: cli/initiator/initiator.go (hunk 9 above)
Comment (pavelkrolevets) [reply cli/initiator/initiator.go]: added HexToAddress with error handling to utils + unit tests for it

File: pkgs/operator/state.go (hunk 11 above)
Comment (pavelkrolevets) [reply pkgs/operator/state.go]: added error wrapps at state. Should we have wrapps everywhere?

**Review by y0sher — APPROVED** [review]

**Comment by lior-blox** [comment]
Merged as part 1 of grooming
